File: swJungle/Week10/algorithm/67258.py
from collections import defaultdict

def solution(gems):
    answer = []
    tmp = 100000000
    target = len(set(gems))
    answer_dict = defaultdict(str)
    
    for i , gem in enumerate(gems):
        answer_dict[gem] = i+1
        
        if len(answer_dict) == target:
            a , b = min(answer_dict.values()), i+1
            if b-a < tmp:
                answer = [a,b]
                tmp = b-a
    
    # print(answer_dict.values())
    return answer

print(solution(["DIA", "RUBY", "RUBY", "DIA", "DIA", "EMERALD", "SAPPHIRE", "DIA"]))
print(solution(["A","A","A","B","B"]))
print(solution(["A"]))
print(solution(["A","B","B","B","B","B","B","C","B","A"]))

File: Algorithm/67258.js
function solution(gems) {
    var answer = [1,gems.length];
    var totalGems = new Set(gems);
    var totalSize = totalGems.size;
    var gemMap = new Map();
    var start = 0;
    var end = 1;
    
    gemMap.set(gems[0], 1);
    
    while(end <= gems.length){
        if(gemMap.size === totalSize){
            // console.log(start,end,"1")
            if(end-1-start < answer[1] - answer[0]){
                // console.log(start, end)
                answer[1] = end;
                answer[0] = start + 1;
            }
            
            var tmp = gemMap.get(gems[start]);
            
            if(tmp === 1){
                gemMap.delete(gems[start]);
            }else{
                gemMap.set(gems[start], tmp -1);
            }
            
            start += 1;
        }else{
            var tmp1 = gemMap.get(gems[end]);
            tmp1 ? gemMap.set(gems[end], tmp1+1) : gemMap.set(gems[end], 1);
            end += 1;
        }
    }
    

    return answer;
}